### PR TITLE
feat(tier4_planning_rviz_plugin): set path colors from rviz and add fade_out feature

### PR DESCRIPTION
**Pull request by KhalilSelyan**
## Description
Updating the style of the path/trajectory display plugin

### Videos

- https://www.youtube.com/playlist?list=PL97OXOB9UA-gSyZWr3VVn4dNnBtt0lqTp

## How was this PR tested?

Just run the planning sim demo.

## Notes for reviewers

We don't need to modify autoware_launch rviz config file because the defaults are set in the plugin.

## Interface changes

New parameters are added:

- property_min_color_("Min Velocity Color", QColor("`#3F2EE3`"), "", &property_path_view_)
- property_mid_color_("Mid Velocity Color", QColor("`#208AAE`"), "", &property_path_view_)
- property_max_color_("Max Velocity Color", QColor("`#00E678`"), "", &property_path_view_)
- property_fade_out_distance_{"Fade Out Distance", 0.0, "[m]", &property_path_view_}

## Effects on system behavior

None.

### CODE REVIEW

**Comment by xmfcx** [comment]
Could you add the following parameter:
**Parameter name:** `fade_out_distance`
**Description:** `The length from the path’s end where it gradually becomes transparent.`

We can discuss the implementation once it is set from rviz

File: common/tier4_planning_rviz_plugin/include/tier4_planning_rviz_plugin/path/display_base.hpp [hunk 1]
```diff
@@ -47,39 +49,6 @@
...
-    double ratio = (cmd_vel_abs - vel_max / 2.0) / (vel_max - vel_max / 2.0);
-    color_ptr = gradation(Qt::yellow, Qt::green, ratio);
-  } else if (vel_max < cmd_vel_abs) {
-    *color_ptr = Ogre::ColourValue::Green;
-  } else {
-    *color_ptr = Ogre::ColourValue::Red;
-  }
-
-  return color_ptr;
-}
```
Comment (xmfcx) [inline common/tier4_planning_rviz_plugin/include/tier4_planning_rviz_plugin/path/display_base.hpp:82]: Why did you move these to protected?
If it's not important, could you move them back to reduce the diff?

File: common/tier4_planning_rviz_plugin/include/tier4_planning_rviz_plugin/path/display_base.hpp [hunk 2]
```diff
@@ -356,23 +415,47 @@ class AutowarePathBaseDisplay : public rviz_common::MessageFilterDisplay<T>
...
+      total_distance += distances[point_idx];
+    }
+
+    double cumulative_distance = 0.0;
+
     for (size_t point_idx = 0; point_idx < msg_ptr->points.size(); point_idx++) {
       const auto & path_point = msg_ptr->points.at(point_idx);
       const auto & pose = autoware::universe_utils::getPose(path_point);
       const auto & velocity = autoware::universe_utils::getLongitudinalVelocity(path_point);
+      // find out the disance between current and start point
```
Comment (xmfcx) [inline common/tier4_planning_rviz_plugin/include/tier4_planning_rviz_plugin/path/display_base.hpp:436]:
```suggestion
      // find out the distance between current and start point
```

File: common/tier4_planning_rviz_plugin/include/tier4_planning_rviz_plugin/path/display_base.hpp (hunk 1 above)
Comment (KhalilSelyan) [reply common/tier4_planning_rviz_plugin/include/tier4_planning_rviz_plugin/path/display_base.hpp:82]: build was complaining about them not having access to properties, I moved them to protected so they can access the class properties like property_min_color_ and still be available for any subclasses that might need them.

File: common/tier4_planning_rviz_plugin/include/tier4_planning_rviz_plugin/path/display_base.hpp [hunk 3]
```diff
@@ -232,6 +201,43 @@ class AutowarePathBaseDisplay : public rviz_common::MessageFilterDisplay<T>
...
+    } else if (vel_max < cmd_vel_abs) {
+      // Use max color when velocity exceeds max
+      *color_ptr = rviz_common::properties::qtToOgre(property_max_color_.getColor());
+    } else {
+      // Use min color when velocity is below min
+      *color_ptr = rviz_common::properties::qtToOgre(property_min_color_.getColor());
+    }
+
+    return color_ptr;
+  }
```
Comment (xmfcx) [inline common/tier4_planning_rviz_plugin/include/tier4_planning_rviz_plugin/path/display_base.hpp:240]: Could you turn:
- `color_min`
- `color_mid`
- `color_max`
into function parameters instead of directly using them here?

Then you can move it back up instead of making it a protected member of the class.

File: common/tier4_planning_rviz_plugin/include/tier4_planning_rviz_plugin/path/display_base.hpp [hunk 4]
```diff
@@ -232,6 +201,43 @@ class AutowarePathBaseDisplay : public rviz_common::MessageFilterDisplay<T>
...
+  {
+    std::unique_ptr<Ogre::ColourValue> color_ptr(new Ogre::ColourValue);
+    color_ptr->g =
+      static_cast<float>(color_max.greenF() * ratio + color_min.greenF() * (1.0 - ratio));
+    color_ptr->r = static_cast<float>(color_max.redF() * ratio + color_min.redF() * (1.0 - ratio));
+    color_ptr->b =
+      static_cast<float>(color_max.blueF() * ratio + color_min.blueF() * (1.0 - ratio));
+
+    return color_ptr;
+  }
```
Comment (xmfcx) [inline common/tier4_planning_rviz_plugin/include/tier4_planning_rviz_plugin/path/display_base.hpp:215]: I think you can move this function back up since they don't use any of the class member variables.

File: common/tier4_planning_rviz_plugin/include/tier4_planning_rviz_plugin/path/display_base.hpp (hunk 1 above)
Comment (xmfcx) [reply common/tier4_planning_rviz_plugin/include/tier4_planning_rviz_plugin/path/display_base.hpp]: Applied with: https://github.com/autowarefoundation/autoware.universe/pull/8972/commits/d8ba16f9a28d27cc156a64f8cd64e15b0e96d58c

File: common/tier4_planning_rviz_plugin/include/tier4_planning_rviz_plugin/path/display_base.hpp (hunk 3 above)
Comment (xmfcx) [reply common/tier4_planning_rviz_plugin/include/tier4_planning_rviz_plugin/path/display_base.hpp]: Applied with: https://github.com/autowarefoundation/autoware.universe/pull/8972/commits/d8ba16f9a28d27cc156a64f8cd64e15b0e96d58c

File: common/tier4_planning_rviz_plugin/include/tier4_planning_rviz_plugin/path/display_base.hpp (hunk 4 above)
Comment (xmfcx) [reply common/tier4_planning_rviz_plugin/include/tier4_planning_rviz_plugin/path/display_base.hpp]: Applied with: https://github.com/autowarefoundation/autoware.universe/pull/8972/commits/d8ba16f9a28d27cc156a64f8cd64e15b0e96d58c

**Comment by xmfcx** [comment]
I think this is ready for review.

@KhalilSelyan could you test the last version?

image

image

image

image

**Comment by xmfcx** [comment]
## Videos

https://www.youtube.com/playlist?list=PL97OXOB9UA-gSyZWr3VVn4dNnBtt0lqTp

**Review by xmfcx — APPROVED** [review]
I've also made changes. So it'd be good if you also reviewed @yukkysaito @takayuki5168

**Review by yukkysaito — APPROVED** [review]
LGTM
